e2e: node: cpumgr: cleanup after each test case

ffromani opened the pull request:
Our CI machines happen to have 1 fully allocatable CPU for test workloads. This is really, really the minimal amount. But still should be sufficient for the tests to run the tests; the CFS quota pod, however, does create a series of pods (at time of writing, 6) and does the cleanup only at the very end the end. This means pods requiring resources accumulate on the CI machine node.

The fix implemented here is to just clean up after each subcase. Doing so the cpu test footprint is equal to the higher requirement (say, 1000 millicores) vs the sum of all the subcases requirements.

Doing like this doesn't change the test behavior, and make it possible to run it on very barebones machines.


#### What type of PR is this?

/kind cleanup
/kind failing-test

#### What this PR does / why we need it:
reduce the CPU footprint of node cpumanager cfs quota testcases, to avoid false negatives reds on CI

#### Which issue(s) this PR fixes:
Fixes #130146

#### Special notes for your reviewer:
N/A

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

ffromani committed:
e2e: node: cpumgr: cleanup after each test case

Our CI machines happen to have 1 fully allocatable CPU for test workloads.
This is really, really the minimal amount. But still should be sufficient for the tests to run
the tests; the CFS quota pod, however, does create a series of pods (at time of writing, 6)
and does the cleanup only at the very end the end. This means pods
requiring resources accumulate on the CI machine node.

The fix implemented here is to just clean up after each subcase.
Doing so the cpu test footprint is equal to the higher requirement (say, 1000 millicores) vs
the sum of all the subcases requirements.

Doing like this doesn't change the test behavior, and make it possible
to run it on very barebones machines.

ffromani commented:
/sig node

ffromani commented:
/cc @kannon92

ffromani commented:
note to self: it would be nice to move towards tabular tests. Here I'm intentionally narrowing down to the minimal (reasonnably clean) fix

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

first thing first: no regressions

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd

unrelated failure

swatisehgal commented on test/e2e_node/cpu_manager_test.go:
```diff
@@ -596,17 +597,30 @@ func runCfsQuotaGuPods(ctx context.Context, f *framework.Framework, disabledCPUQ
-	var cleanupPods []*v1.Pod
+	podsToClean := make(map[apitypes.UID]*v1.Pod) // pod.UID -> pod
+
+	cleanupTestPod := func(ctx2 context.Context, pod *v1.Pod) {
```
Why restrict its scope to only this function? We could potentially use this in other places as well.

ffromani replied:
right, it's a pattern we have. Added another helper.

swatisehgal commented on test/e2e_node/cpu_manager_test.go:
```diff
@@ -596,17 +597,30 @@ func runCfsQuotaGuPods(ctx context.Context, f *framework.Framework, disabledCPUQ
+		framework.Logf("deleted pod: %s/%s", pod.Namespace, pod.Name)
+	}
+
+	deleteTestPod := func(pod *v1.Pod) {
+		// waitForContainerRemoval takes "long" to complete; if we use the parent ctx we get a
+		// 'deadline expired' message and the cleanup aborts, which we don't want.
```
Same as above?

ffromani replied:
this one wants to be a closure thought because it manages the `podsToClean` map. I think most of the value comes from the previous helper anyway.

swatisehgal commented on test/e2e_node/cpu_manager_test.go:
```diff
@@ -596,17 +597,30 @@ func runCfsQuotaGuPods(ctx context.Context, f *framework.Framework, disabledCPUQ
+	deleteTestPod := func(pod *v1.Pod) {
+		// waitForContainerRemoval takes "long" to complete; if we use the parent ctx we get a
+		// 'deadline expired' message and the cleanup aborts, which we don't want.
+		cleanupTestPod(context.TODO(), pod)
```
What if there are failures in cleanup? With `context.TODO()` we could end up in indefinite waits in case of failures so perhaps it is better to use ` context .WithTImeout `with some reasonable timeout?

ffromani replied:
good point, waiting for hours just to see the lane fail is bad dev UX anyway. Adding a timeout.

swatisehgal reviewed (commented):
Overall LGTM

Left some minor comments/questions.

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test

ffromani commented:
/test pull-kubernetes-e2e-ec2
/test pull-kubernetes-e2e-ec2-conformance
/test pull-kubernetes-node-e2e-containerd-ec2
/test pull-kubernetes-node-e2e-containerd-features
/test pull-kubernetes-node-e2e-containerd-features-kubetest2
/test pull-kubernetes-node-e2e-containerd-kubetest2
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-e2e-containerd-serial-ec2-eks

ffromani commented:
/test pull-crio-cgroupv2-node-e2e-hugepages
/test pull-crio-cgroupv2-node-e2e-hugepages-kubetest2
/test pull-crio-cgroupv2-node-e2e-resource-managers
/test pull-crio-cgroupv2-node-e2e-resource-managers-kubetest2
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

ffromani commented:
/test pull-kubernetes-node-crio-cgrpv2-e2e
/test pull-kubernetes-node-crio-cgrpv2-e2e-kubetest2
/test pull-kubernetes-node-crio-cgrpv2-userns-e2e-serial
/test pull-kubernetes-node-crio-cgrpv2-userns-e2e-serial-kubetest2
/test pull-kubernetes-node-crio-e2e
/test pull-kubernetes-node-crio-e2e-kubetest2

ffromani commented:
note to self: https://github.com/kubernetes/kubernetes/pull/127525#issuecomment-2606950229

ffromani commented:
>pull-kubernetes-node-crio-cgrpv2-e2e

ffromani commented:
the failure of the `pull-kubernetes-node-e2e-containerd-serial-ec2` lane seems real:
```
I0214 16:33:04.497614 26109 dump.go:53] At 2025-02-14 16:32:57 +0000 UTC - event for gu-pod4: {kubelet i-0c3cd180f6bd72e49} OutOfcpu: Node didn't have enough resource: cpu, requested: 1500, used: 0, capacity: 1000
```
more changes are needed

ffromani committed:
e2e: node: cpumgr: check CPU allocatable for CFS quota test

add (admittedly pretty crude) CPU allocatable check.
A more incisive refactoring is needed, but we need
to unbreak CI first, so this seems the minimal decently clean test.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2

ffromani commented:
/test pull-kubernetes-e2e-ec2
/test pull-kubernetes-e2e-ec2-conformance
/test pull-kubernetes-node-e2e-containerd-ec2
/test pull-kubernetes-node-e2e-containerd-features
/test pull-kubernetes-node-e2e-containerd-features-kubetest2
/test pull-kubernetes-node-e2e-containerd-kubetest2
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-e2e-containerd-serial-ec2-eks
/test pull-crio-cgroupv2-node-e2e-hugepages
/test pull-crio-cgroupv2-node-e2e-hugepages-kubetest2
/test pull-crio-cgroupv2-node-e2e-resource-managers
/test pull-crio-cgroupv2-node-e2e-resource-managers-kubetest2
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-crio-cgroupv2-node-e2e-hugepages
/test pull-crio-cgroupv2-node-e2e-hugepages-kubetest2
/test pull-crio-cgroupv2-node-e2e-resource-managers
/test pull-crio-cgroupv2-node-e2e-resource-managers-kubetest2
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/retest-required

the lane to focus on in the short term is `/test pull-kubernetes-node-e2e-containerd-serial-ec2`

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2

ffromani commented:
/retest-required

```
Kubernetes e2e suite: [It] [sig-cli] Kubectl client Simple pod should contain last line of the log 
```

ffromani commented:
/test pull-kubernetes-integration

unrelated

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2

silly mistake, missing to include a case in the guard

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-e2e-ec2
/test pull-kubernetes-e2e-ec2-conformance
/test pull-kubernetes-node-e2e-containerd-ec2
/test pull-kubernetes-node-e2e-containerd-features
/test pull-kubernetes-node-e2e-containerd-features-kubetest2
/test pull-kubernetes-node-e2e-containerd-kubetest2
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-e2e-containerd-serial-ec2-eks
/test pull-crio-cgroupv2-node-e2e-hugepages
/test pull-crio-cgroupv2-node-e2e-hugepages-kubetest2
/test pull-crio-cgroupv2-node-e2e-resource-managers
/test pull-crio-cgroupv2-node-e2e-resource-managers-kubetest2
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-crio-cgroupv2-node-e2e-hugepages
/test pull-crio-cgroupv2-node-e2e-hugepages-kubetest2
/test pull-crio-cgroupv2-node-e2e-resource-managers
/test pull-crio-cgroupv2-node-e2e-resource-managers-kubetest2
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-features-kubetest2

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-crio-cgrpv2-e2e

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-crio-cgrpv2-e2e

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2

ffromani commented:
we're close to the final fix:
```
I0217 16:34:03.822014   23392 cpu_manager.go:178] "Detected CPU topology" topology={"NumCPUs":2,"NumCores":1,"NumUncoreCache":1,"NumSockets":1,"NumNUMANodes":1,"CPUDetails":{"0":{"NUMANodeID":0,"SocketID":0,"CoreID":0,"UncoreCacheID":0},"1":{"NUMANodeID":0,"SocketID":0,"CoreID":0,"UncoreCacheID":0}}}
I0217 16:34:03.822088   23392 policy_static.go:143] "Static policy created with configuration" options={"FullPhysicalCPUsOnly":false,"DistributeCPUsAcrossNUMA":false,"AlignBySocket":false,"DistributeCPUsAcrossCores":false,"StrictCPUReservation":false,"PreferAlignByUncoreCacheOption":false} cpuGroupSize=2
I0217 16:34:03.822134   23392 policy_static.go:180] "Reserved CPUs not available for exclusive assignment" reservedSize=1 reserved="0" reservedPhysicalCPUs="0-1"
```

so yes, CI machines have 2 cpus, and our detection logic in the test is still at fault

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-serial-ec2

ffromani commented:
the latest upload eventually fixed the cpumanager quota failures:
![image](https://github.com/user-attachments/assets/a3dbde78-8aa0-43a5-b3f2-66e3df3e33c4)

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-e2e-ec2
/test pull-kubernetes-e2e-ec2-conformance
/test pull-kubernetes-node-e2e-containerd-ec2
/test pull-kubernetes-node-e2e-containerd-features
/test pull-kubernetes-node-e2e-containerd-features-kubetest2
/test pull-kubernetes-node-e2e-containerd-kubetest2
/test pull-kubernetes-node-e2e-containerd-serial-ec2
/test pull-kubernetes-node-e2e-containerd-serial-ec2-eks
/test pull-crio-cgroupv2-node-e2e-hugepages
/test pull-crio-cgroupv2-node-e2e-hugepages-kubetest2
/test pull-crio-cgroupv2-node-e2e-resource-managers
/test pull-crio-cgroupv2-node-e2e-resource-managers-kubetest2
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-crio-cgroupv2-node-e2e-hugepages
/test pull-crio-cgroupv2-node-e2e-hugepages-kubetest2
/test pull-crio-cgroupv2-node-e2e-resource-managers
/test pull-crio-cgroupv2-node-e2e-resource-managers-kubetest2
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

ffromani commented:
none of the failures are related to cfs quota

ffromani commented:
/test pull-kubernetes-node-e2e-containerd-features

ffromani commented:
@swatisehgal (cc @kannon92 ) ready for review!

swatisehgal reviewed (commented):
/lgtm

I am happy enough to proceed with this and would like to unblock the CI. There is some room for improvement as has been already captured in the PR but I am okay to defer that to a later stage.

/hold
to allow reviews from other reviewers

bart0sh commented:
/test pull-crio-cgroupv1-node-e2e-resource-managers
/test pull-crio-cgroupv2-node-e2e-resource-managers
/test pull-crio-cgroupv1-node-e2e-resource-managers-kubetest2
/test pull-crio-cgroupv2-node-e2e-resource-managers-kubetest2

bart0sh commented:
/lgtm
/retest

ffromani commented:
hey folks, thanks for the reviews. Any remaining concerns or comments? otherwise I think we should remove the hold.

bart0sh commented:
/retest
